### PR TITLE
Upgrades moto to newer version (~=2.0)

### PR DESCRIPTION
**Pull request by potiuk**
According to https://github.com/spulec/moto/issues/3535#issuecomment-808706939
1.3.17 version of moto with a fix to be compatible with mock> 4.0.3 is
not going to be released because of breaking changes. Therefore we need
to migrate to newer version of moto.

At the same time we can get rid of the old botocore limitation, which
was added apparently to handle some test errors. We are relying fully
on what boto3 depends on.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

**Comment by kaxil** [comment]
This should close https://github.com/apache/airflow/issues/14434

**Review by kaxil — APPROVED** [review]

File: tests/providers/amazon/aws/hooks/test_secrets_manager.py [hunk 1]
```diff
@@ -43,12 +43,17 @@ def test_get_secret_string(self):
...
+            'Name': secret_name,
+        }
+
+        put_param = {
             'SecretId': secret_name,
             'SecretString': secret_value,
         }
 
-        hook.get_conn().put_secret_value(**param)
+        hook.get_conn().create_secret(**create_param)
```
Comment (bblommers) [inline tests/providers/amazon/aws/hooks/test_secrets_manager.py:55]: Background info on why this change is necessary: AWS will throw an error if you call `put_secret_value` on a secret that hasn't been explicitly created first. Moto now follows that behaviour, as of 2.x.

If you want to simplify these tests: there's no need to both create and update the secret. It works by just creating the secret:
```
secret_name = "db_cluster-YYYYYYY"  # no need for this to be an ARN (anymore)
param = {
            'Name': secret_name,
            'SecretString': secret_value_binary,
        }

        hook.get_conn().create_secret(**param)
```
The three tests all pass on moto>=2 with this pattern.

Comment (potiuk) [reply tests/providers/amazon/aws/hooks/test_secrets_manager.py:55]: Nice. But the whole point of those tests is to test "put_secret_value" method really, not the "create_secret" one :). If we simplify it, we will test different things.

Comment (bblommers) [reply tests/providers/amazon/aws/hooks/test_secrets_manager.py:55]: Gotcha, my bad. My assumption was that the point was to test the creation/retrieval flow in general.
